torch/csrc/distributed/rpc/testing/faulty_process_group_agent.cpp

```diff
@@ -98,6 +101,16 @@ void FaultyProcessGroupAgent::enqueueSend(SendWork work) {
   ProcessGroupAgent::enqueueSend(std::move(work));
 }
 
+void FaultyProcessGroupAgent::sendToSelf(Message&& message) {
+  float msgDelay = getDelayForMessage(message.type());
+  if (msgDelay != 0) {
+    // Sleep for the specified delay for the message.
+    std::this_thread::sleep_for(std::chrono::milliseconds(
```

Comment:
Just a thought: Maybe to make the sleep action to be slightly more realistic (i.e. only affect the specified message, vs pausing everything behind it as well), it might be worth spawning a detached thread?, e.g.

```
 std::thread thr([msgDelay, message, this]() {
   std::this_thread::sleep_for(std::chrono::milliseconds(...);
   ProcessGroupAgent::sendToSelf(std::move(message));
  });
  thr.detach();
```

The detached thread is sketchy in prod code, seems reasonable for test cases

Reply:
Great idea! This will definitely make the delays more realistic.

Reply:
I will plan to do this in a follow up PR.